lib/migrate-existing-resources.js

```diff
@@ -13,15 +15,20 @@ module.exports = function getCurrentState() {
     .then(stackSummary => stackSummary.filter(summary => summary.ResourceType === 'AWS::CloudFormation::Stack'))
     .then(nestedStacks => {
       return Promise.all(nestedStacks.map(stack => {
-        const nestedStackName = stack.LogicalResourceId;
         const nestedStackId = stack.PhysicalResourceId;
 
         return this.getStackSummary(nestedStackId)
           .then(nestedStackSummary => {
             nestedStackSummary.forEach(nestedSummary => {
-              if (nestedSummary.LogicalResourceId in this.resourcesById) {
-                this.migrate(nestedSummary.LogicalResourceId, nestedStackName);
-              }
+              const logicalId = nestedSummary.LogicalResourceId;
+              const resource = this.resourcesById[logicalId];
+              if (!resource) return;
+              const migration = migrations[resource.Type];
+              if (!migration) return;
+              const destination = (typeof migration.destination === 'function')
+                ? migration.destination(logicalId, resource) : migration.destination;
```

Comment:
should we pass a third parameter here (the existing stack) so dest functions can optionally keep things in the stack they already went to?

Reply:
We may, although I was wondering about purpose of this step (_migrate-existing-resources_).

Technically complete source of information about resources before we start any mangling should be at `this.serverless.service.provider.compiledCloudFormationTemplate`.

If it's the case why do we bother reading what's already deployed to CF? We shouldn't care. If given resource is not anymore in `compiledCloudFormationTemplate` then it means it was removed. If it is then it'll be propagated to accurate stack with _migrate-new-resources_ step.

Other, probably most important point, is that handling of CF stacks in Serverless is in some sense __stateless__. Serverless (as far as I understand how it works), does not investigate what was deployed before, and it's quite valuable characteristics (not relying on current state makes things significantly simpler).  Here we're sort of breaking it, by affecting outcome on basis of what was deployed before.

Reply:
I had to add the fetching current state because you may be migrated more resources than can go in a sub-stack -- example, 201 cloudwatch alarms. Without fetching the current state, you're at the mercy of the iteration order of the template JSON, and it means you can have resources jump from one stack to the other, but CF blows up because it doesn't understand "moving" between nested stacks (it just sees a create in one stack, and tries to process it, but the delete in the other stack did not occur yet).

It's a similar case to the first time you deploy with this plugin, where a resource is being migrated from the root stack in to a nested one, but CF thinks it still exists. (I've found no way of guaranteeing a zero downtime switch from not using this plugin, to using this plugin)

Reply:
Ok, I got it, so _merge-existing-resources_ ensures that those which already were deployed to given nested stack are treated with priority, so there's no case they eventually  end in the other stack (due to resources overflow). Makes sense.

I've updated the changes, as it didn't cover that idea as expected. I've also added additional comments to clarify purpose of  _merge-existing-resources_

Now, I think we should not provide there existing stack name to _destination_ function, as it should not influence in any way `destination` result (it will stop being deterministic if it will).

If I understand right eventual stack suffixing is handled outside of _destination_ scope, and technically result of `destination` never matches exactly the stack name (it's suffixed with `NestedStack`, and with eventual number suffix if `allowSuffix` is set on migration configuration)

Reply:
You're right - good call.

Yeah we have services that overflow their nested stacks and the overflow isn't predictable because one serverless plugin might add 5 resources *per function* and so when you add a new function to your service it causes resources added by a second plugin that are added later to suddenly overflow. It's not predictable.

Underflow also doesn't work - you can't reliably 'compact' resources back in to a single nested stack even if theres room. Sometimes the deploy just fails.